### ISSUE TITLE
Lua/Redis: BUSY Redis is busy running a script

### ISSUE THREAD

**Issue by ksheurs** [jruizgit/rules#108]
Hi @jruizgit! I'm close with my project all working. I've started to feed the rules more production-based data and I'm seeing as I send in more facts to be evaluated, I start to see script timeout/busy issues.

Some example exceptions:

```
getReply err 1, 35, Resource temporarily unavailable
peekTimers err string BUSY Redis is busy running a script. You can only call SCRIPT KILL or SHUTDOWN NOSAVE.
unknown exception Could not complete and start action, error code: 306
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:715:in `complete_and_start_action'
unexpected error Could not assert timers, error code: 302/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:715:in `block (2 levels) in dispatch'

/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:710:in `each'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:710:in `block in dispatch'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:340:in `run'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:338:in `run'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:625:in `dispatch'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:1304:in `block in start_dispatch_ruleset_thread'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/timer.rb:97:in `fire'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/events.rb:43:in `fire'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/events.rb:86:in `block in fire'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/events.rb:85:in `reverse_each'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/events.rb:85:in `fire'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/group.rb:94:in `fire'
/usr/local/lib/ruby/gems/2.4.0/gems/timers-4.0.4/lib/timers/group.rb:78:in `wait'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:1314:in `block (2 levels) in start_dispatch_ruleset_thread'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:1314:in `loop'
/usr/local/lib/ruby/gems/2.4.0/gems/durable_rules-0.34.39/librb/engine.rb:1314:in `block in start_dispatch_ruleset_thread
```

If I set `redis-cli config set lua-time-limit 0`, a bulk of these errors go away, but I still see `getReply err 1, 35, Resource temporarily unavailable` and a good amount of hanging/delays (~15s) before the burst of expected output.

You can put these 3 files in the same directory and run the bash script to reproduce: https://gist.github.com/ksheurs/38d33237ddf6bceb77526bd5b3c833dd. 

Any insights into what's going on here? Thanks!

**Comment by jruizgit** [comment]
That shouldn't happen. Let me debug your scenario. I will report back on my findings later this week.

**Comment by jruizgit** [comment]
Hi, I played with this scenario for a little while. By the time the `try_assignment` state is reached, this rule as accumulated 32k possible solutions. Is there a way you can make the rule more specific?

```ruby
    # Remaining / unlinked, crops (length 2) into trays
    to :tray_assignment, when_all(pri(4), c.tray = (m.t == "tray") & (m.raft_count <= 2),
                                          c.r1   = (m.t == "raft"),
                                          c.r2   = (m.t == "raft") & (m.raft_id != r1.raft_id),
                                          none((m.t == "link") & (m.r1 == r1.raft_id) & (m.r2 == r2.raft_id)),
                                          none((m.t == "link") & (m.r1 == r2.raft_id) & (m.r2 == r1.raft_id)),
                                          none((m.t == "assignment") & (m.raft_id == r1.raft_id)),
                                          none((m.t == "assignment") & (m.raft_id == r2.raft_id))) do
      assign_tray(s, tray, r1, r2)
    end
```

**Comment by ksheurs** [comment]
Oh wow, ok. Thanks for debugging! I've since made that rule more specific and the errors have seemed to go away.

Do you see a way where you could run durable in a debug or verbose mode? It'd be great to see these kind of insights or which rules are trying to be satisfied against which facts.
